core/src/main/java/org/apache/calcite/adapter/enumerable/PhysTypeImpl.java

```diff
@@ -582,7 +590,8 @@ public Expression generateAccessor(
       // }
       Expressions.FluentList<Expression> list = Expressions.list();
       for (int field : fields) {
-        list.add(fieldReference(v1, field));
+        list.add(
+            Expressions.call(CONVERT2STRING_METHOD, fieldReference(v1, field)));
       }
```

Comment:
Need test case at least. Besides, I really don't think it's a good idea to convert join keys to strings since its costs.

Reply:
@chunweilei ，I pushed a new commit , can you review it?